cudax/include/cuda/experimental/__execution/bulk.cuh

```diff
@@ -0,0 +1,479 @@
+                        "The default implementation cannot provide the concurrent progress guarantee");
+        }
+        else
+        {
+          const auto __guarantee =
+            get_forward_progress_guarantee(get_completion_scheduler<set_value_t>(execution::get_env(__sndr)));
+          _CCCL_ASSERT(__guarantee != forward_progress_guarantee::concurrent,
+                       "The default implementation cannot provide the concurrent progress guarantee");
+        }
+      }
+    }
+
+    _CCCL_EXEC_CHECK_DISABLE
+    template <class... _Values>
+    _CCCL_API void set_value(_Values&&... __values) noexcept
+    {
+      _CUDAX_TRY( //
+        ({
+          for (_Shape __index{}; __index != this->__shape_; ++__index)
+          {
+            this->__fn_(_Shape(__index), __values...);
+          }
+          execution::set_value(static_cast<_Rcvr&&>(this->__rcvr_), static_cast<_Values&&>(__values)...);
+        }),
+        _CUDAX_CATCH(...) //
+        ({
+          if constexpr (!__nothrow_callable<_Fn&, _Shape, _Values&...>)
+          {
+            execution::set_error(static_cast<_Rcvr&&>(this->__rcvr_), ::std::current_exception());
+          }
+        }))
+    }
+  };
+
+  // The bulk_unchunked algorithm is different from the other bulk algorithms in that it
+  // does not take an execution policy, so we must hide the base class's `operator()` that
+  // takes a policy with an overload that does not take a policy.
+  template <class _Sndr, class _Shape, class _Fn>
+  [[nodiscard]] _CCCL_TRIVIAL_API constexpr auto operator()(_Sndr&& __sndr, _Shape __shape, _Fn __fn) const
+  {
+    return this->__bulk_t::operator()(static_cast<_Sndr&&>(__sndr), par, __shape, static_cast<_Fn&&>(__fn));
+  }
+
+  template <class _Shape, class _Fn>
+  [[nodiscard]] _CCCL_TRIVIAL_API constexpr auto operator()(_Shape __shape, _Fn __fn) const
+  {
+    return this->__bulk_t::operator()(par, __shape, static_cast<_Fn&&>(__fn));
+  }
+
+  _CCCL_API static constexpr bool __is_chunked() noexcept
+  {
+    return false;
+  }
+};
+
+_CCCL_GLOBAL_CONSTANT auto bulk_unchunked = bulk_unchunked_t{};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+// bulk
+struct _CCCL_TYPE_VISIBILITY_DEFAULT bulk_t : __bulk_t<bulk_t>
+{
+  // This is a function adaptor that transforms a `bulk` function that takes a single
+  // shape to a `bulk_chunked` function that takes a begin and end shape.
+  template <class _Shape, class _Fn>
+  struct __bulk_chunked_fn
+  {
+    _CCCL_EXEC_CHECK_DISABLE
+    template <class... _Ts>
+    _CCCL_TRIVIAL_API auto operator()(_Shape __begin, _Shape __end, _Ts&&... __values) noexcept(
+      __nothrow_callable<_Fn&, _Shape, decltype(__values)&...>)
+    {
+      for (; __begin != __end; ++__begin)
+      {
+        // Pass a copy of `__begin` to the function so it can't do anything funny with it.
+        __fn_(_Shape(__begin), __values...);
```

Comment:
Question: Should this use `invoke`?

Reply:
`invoke` only does special things for member pointers, and in that case, the second argument must be an object to which the member pointer can be applied. here, the second argument is an integer, so `invoke` would do nothing different than this.